File: sandbox/protocol/game.ts
/**
 * Game Protocol Packets
 * 
 * Core gameplay structures: towers, battles, harvests, achievements, etc.
 */

import { BinaryReader, BinaryWriter } from "./binary";
import { CCommonRespMsg, writeCCommonRespMsg, createSuccessResponse } from "./common";

// =============================================================================
// GAME TOWER
// =============================================================================

export interface CGameTowerInfo {
  m_nType: number;              // UInt16
  m_bWin: boolean;              // Boolean
  m_nTransID: number;           // UInt32
}

export function readCGameTowerInfo(reader: BinaryReader): CGameTowerInfo {
  return {
    m_nType: reader.readUInt16(),
    m_bWin: reader.readBool(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface CPlayTowerInfo {
  m_nType: number;              // UInt16
  m_nTowerId: number;           // UInt32
  m_nFloor: number;             // UInt16
  m_nTransID: number;           // UInt32
}

export function readCPlayTowerInfo(reader: BinaryReader): CPlayTowerInfo {
  return {
    m_nType: reader.readUInt16(),
    m_nTowerId: reader.readUInt32(),
    m_nFloor: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface CRespGameTowerInfo {
  m_stRetMsg: CCommonRespMsg;
  m_nFloor: number;             // UInt16
  m_nMaxFloor: number;          // UInt16
}

export function writeCRespGameTowerInfo(writer: BinaryWriter, resp: CRespGameTowerInfo): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt16(resp.m_nFloor);
  writer.writeUInt16(resp.m_nMaxFloor);
}

export interface CRespPlayTowerInfo {
  m_stRetMsg: CCommonRespMsg;
  m_nTowerId: number;           // UInt32
  m_nFloor: number;             // UInt16
}

export function writeCRespPlayTowerInfo(writer: BinaryWriter, resp: CRespPlayTowerInfo): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nTowerId);
  writer.writeUInt16(resp.m_nFloor);
}

// =============================================================================
// HARVEST
// =============================================================================

export interface CReqGameHarvest2 {
  m_nType: number;              // UInt16
  m_nTransID: number;           // UInt32
}

export function readCReqGameHarvest2(reader: BinaryReader): CReqGameHarvest2 {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface CRespGameHarvest2 {
  m_stRetMsg: CCommonRespMsg;
  m_nCoins: number;             // UInt32
  m_nExp: number;               // UInt32
  m_nTimestamp: bigint;         // UInt64
  m_nMaxTime: number;           // UInt32
}

export function writeCRespGameHarvest2(writer: BinaryWriter, resp: CRespGameHarvest2): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nCoins);
  writer.writeUInt32(resp.m_nExp);
  writer.writeUInt64(resp.m_nTimestamp);
  writer.writeUInt32(resp.m_nMaxTime);
}

// =============================================================================
// ACHIEVEMENTS
// =============================================================================

export interface CGameAchieveInfo {
  m_nType: number;              // UInt16
  m_nId: number;                // UInt32
  m_nTransID: number;           // UInt32
}

export function readCGameAchieveInfo(reader: BinaryReader): CGameAchieveInfo {
  return {
    m_nType: reader.readUInt16(),
    m_nId: reader.readUInt32(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface STCommonAchievementData {
  m_nId: number;                // UInt32
  m_nProgress: number;          // UInt32
  m_nLevel: number;             // UInt16
  m_bIsClaimed: boolean;        // Boolean
}

export function writeSTCommonAchievementData(writer: BinaryWriter, data: STCommonAchievementData): void {
  writer.writeUInt32(data.m_nId);
  writer.writeUInt32(data.m_nProgress);
  writer.writeUInt16(data.m_nLevel);
  writer.writeBool(data.m_bIsClaimed);
}

export interface CRespGameAchieveInfo {
  m_stRetMsg: CCommonRespMsg;
  m_vecAchievements: STCommonAchievementData[];
}

export function writeCRespGameAchieveInfo(writer: BinaryWriter, resp: CRespGameAchieveInfo): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeArray(resp.m_vecAchievements, (a) => writeSTCommonAchievementData(writer, a));
}

// =============================================================================
// ADS
// =============================================================================

export interface CGameAd {
  m_nType: number;              // UInt16
  m_nAdId: number;              // UInt32
  m_nTransID: number;           // UInt32
}

export function readCGameAd(reader: BinaryReader): CGameAd {
  return {
    m_nType: reader.readUInt16(),
    m_nAdId: reader.readUInt32(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface CRespGameAd {
  m_stRetMsg: CCommonRespMsg;
  m_nAdCount: number;           // UInt16
  m_nDailyLimit: number;        // UInt16
}

export function writeCRespGameAd(writer: BinaryWriter, resp: CRespGameAd): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt16(resp.m_nAdCount);
  writer.writeUInt16(resp.m_nDailyLimit);
}

// =============================================================================
// GUIDE
// =============================================================================

export interface CReqGameGuide {
  m_nType: number;              // UInt16
  m_nGuideId: number;           // UInt32
  m_nTransID: number;           // UInt32
}

export function readCReqGameGuide(reader: BinaryReader): CReqGameGuide {
  return {
    m_nType: reader.readUInt16(),
    m_nGuideId: reader.readUInt32(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface CRespGameGuide {
  m_stRetMsg: CCommonRespMsg;
  m_nGuideBits: bigint;         // UInt64
}

export function writeCRespGameGuide(writer: BinaryWriter, resp: CRespGameGuide): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt64(resp.m_nGuideBits);
}

// =============================================================================
// GAME CLIENT DATA
// =============================================================================

export interface CReqGameClientData {
  m_nType: number;              // UInt16
  m_strClientData: string | null; // String
}

export function readCReqGameClientData(reader: BinaryReader): CReqGameClientData {
  return {
    m_nType: reader.readUInt16(),
    m_strClientData: reader.readString(),
  };
}

export interface CRespGameClientData {
  m_stRetMsg: CCommonRespMsg;
  m_strClientData: string;      // String
}

export function writeCRespGameClientData(writer: BinaryWriter, resp: CRespGameClientData): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeString(resp.m_strClientData);
}

// =============================================================================
// PVE SEASON
// =============================================================================

export interface CPveSeasonInfo {
  m_nType: number;              // UInt16
  m_nSeasonId: number;          // UInt32
  m_nTransID: number;           // UInt32
}

export function readCPveSeasonInfo(reader: BinaryReader): CPveSeasonInfo {
  return {
    m_nType: reader.readUInt16(),
    m_nSeasonId: reader.readUInt32(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface CRespPveSeasonInfo {
  m_stRetMsg: CCommonRespMsg;
  m_nSeasonId: number;          // UInt32
  m_nRank: number;              // UInt32
  m_nScore: number;             // UInt32
}

export function writeCRespPveSeasonInfo(writer: BinaryWriter, resp: CRespPveSeasonInfo): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nSeasonId);
  writer.writeUInt32(resp.m_nRank);
  writer.writeUInt32(resp.m_nScore);
}

// =============================================================================
// FISHING
// =============================================================================

export interface CReqGameFishing {
  m_nType: number;              // UInt16
  m_nTransID: number;           // UInt32
}

export function readCReqGameFishing(reader: BinaryReader): CReqGameFishing {
  return {
    m_nType: reader.readUInt16(),
    m_nTransID: reader.readUInt32(),
  };
}

export interface STGameFishingRank {
  m_nRank: number;              // UInt32
  m_nScore: number;             // UInt32
  m_strName: string;            // String
}

export function writeSTGameFishingRank(writer: BinaryWriter, rank: STGameFishingRank): void {
  writer.writeUInt32(rank.m_nRank);
  writer.writeUInt32(rank.m_nScore);
  writer.writeString(rank.m_strName);
}

export interface CRespGameFishing {
  m_stRetMsg: CCommonRespMsg;
  m_nScore: number;             // UInt32
  m_vecRanks: STGameFishingRank[];
}

export function writeCRespGameFishing(writer: BinaryWriter, resp: CRespGameFishing): void {
  writeCCommonRespMsg(writer, resp.m_stRetMsg);
  writer.writeUInt32(resp.m_nScore);
  writer.writeArray(resp.m_vecRanks, (r) => writeSTGameFishingRank(writer, r));
}

File: server-py/protocol/game.py
"""
Game Protocol Packets

Core gameplay structures: towers, battles, harvests, achievements, etc.
"""

from dataclasses import dataclass, field
from typing import List, Optional

from .binary import BinaryReader, BinaryWriter
from .common import CCommonRespMsg, write_c_common_resp_msg


# =============================================================================
# GAME TOWER
# =============================================================================


@dataclass
class CGameTowerInfo:
    m_nType: int  # UInt16
    m_bWin: bool  # Boolean
    m_nTransID: int  # UInt32


def read_c_game_tower_info(reader: BinaryReader) -> CGameTowerInfo:
    return CGameTowerInfo(
        m_nType=reader.read_uint16(),
        m_bWin=reader.read_bool(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class CPlayTowerInfo:
    m_nType: int  # UInt16
    m_nTowerId: int  # UInt32
    m_nFloor: int  # UInt16
    m_nTransID: int  # UInt32


def read_c_play_tower_info(reader: BinaryReader) -> CPlayTowerInfo:
    return CPlayTowerInfo(
        m_nType=reader.read_uint16(),
        m_nTowerId=reader.read_uint32(),
        m_nFloor=reader.read_uint16(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class CRespGameTowerInfo:
    m_stRetMsg: CCommonRespMsg
    m_nFloor: int = 0  # UInt16
    m_nMaxFloor: int = 0  # UInt16


def write_c_resp_game_tower_info(
    writer: BinaryWriter, resp: CRespGameTowerInfo
) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint16(resp.m_nFloor)
    writer.write_uint16(resp.m_nMaxFloor)


@dataclass
class CRespPlayTowerInfo:
    m_stRetMsg: CCommonRespMsg
    m_nTowerId: int = 0  # UInt32
    m_nFloor: int = 0  # UInt16


def write_c_resp_play_tower_info(
    writer: BinaryWriter, resp: CRespPlayTowerInfo
) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint32(resp.m_nTowerId)
    writer.write_uint16(resp.m_nFloor)


# =============================================================================
# HARVEST
# =============================================================================


@dataclass
class CReqGameHarvest2:
    m_nType: int  # UInt16
    m_nTransID: int  # UInt32


def read_c_req_game_harvest2(reader: BinaryReader) -> CReqGameHarvest2:
    return CReqGameHarvest2(
        m_nType=reader.read_uint16(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class CRespGameHarvest2:
    m_stRetMsg: CCommonRespMsg
    m_nCoins: int = 0  # UInt32
    m_nExp: int = 0  # UInt32
    m_nTimestamp: int = 0  # UInt64
    m_nMaxTime: int = 0  # UInt32


def write_c_resp_game_harvest2(writer: BinaryWriter, resp: CRespGameHarvest2) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint32(resp.m_nCoins)
    writer.write_uint32(resp.m_nExp)
    writer.write_uint64(resp.m_nTimestamp)
    writer.write_uint32(resp.m_nMaxTime)


# =============================================================================
# ACHIEVEMENTS
# =============================================================================


@dataclass
class CGameAchieveInfo:
    m_nType: int  # UInt16
    m_nId: int  # UInt32
    m_nTransID: int  # UInt32


def read_c_game_achieve_info(reader: BinaryReader) -> CGameAchieveInfo:
    return CGameAchieveInfo(
        m_nType=reader.read_uint16(),
        m_nId=reader.read_uint32(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class STCommonAchievementData:
    m_nId: int  # UInt32
    m_nProgress: int  # UInt32
    m_nLevel: int  # UInt16
    m_bIsClaimed: bool  # Boolean


def write_st_common_achievement_data(
    writer: BinaryWriter, data: STCommonAchievementData
) -> None:
    writer.write_uint32(data.m_nId)
    writer.write_uint32(data.m_nProgress)
    writer.write_uint16(data.m_nLevel)
    writer.write_bool(data.m_bIsClaimed)


@dataclass
class CRespGameAchieveInfo:
    m_stRetMsg: CCommonRespMsg
    m_vecAchievements: List[STCommonAchievementData] = field(default_factory=list)


def write_c_resp_game_achieve_info(
    writer: BinaryWriter, resp: CRespGameAchieveInfo
) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_array(
        resp.m_vecAchievements, lambda a: write_st_common_achievement_data(writer, a)
    )


# =============================================================================
# ADS
# =============================================================================


@dataclass
class CGameAd:
    m_nType: int  # UInt16
    m_nAdId: int  # UInt32
    m_nTransID: int  # UInt32


def read_c_game_ad(reader: BinaryReader) -> CGameAd:
    return CGameAd(
        m_nType=reader.read_uint16(),
        m_nAdId=reader.read_uint32(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class CRespGameAd:
    m_stRetMsg: CCommonRespMsg
    m_nAdCount: int = 0  # UInt16
    m_nDailyLimit: int = 0  # UInt16


def write_c_resp_game_ad(writer: BinaryWriter, resp: CRespGameAd) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint16(resp.m_nAdCount)
    writer.write_uint16(resp.m_nDailyLimit)


# =============================================================================
# GUIDE
# =============================================================================


@dataclass
class CReqGameGuide:
    m_nType: int  # UInt16
    m_nGuideId: int  # UInt32
    m_nTransID: int  # UInt32


def read_c_req_game_guide(reader: BinaryReader) -> CReqGameGuide:
    return CReqGameGuide(
        m_nType=reader.read_uint16(),
        m_nGuideId=reader.read_uint32(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class CRespGameGuide:
    m_stRetMsg: CCommonRespMsg
    m_nGuideBits: int = 0  # UInt64


def write_c_resp_game_guide(writer: BinaryWriter, resp: CRespGameGuide) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint64(resp.m_nGuideBits)


# =============================================================================
# GAME CLIENT DATA
# =============================================================================


@dataclass
class CReqGameClientData:
    m_nType: int  # UInt16
    m_strClientData: Optional[str] = None  # String


def read_c_req_game_client_data(reader: BinaryReader) -> CReqGameClientData:
    return CReqGameClientData(
        m_nType=reader.read_uint16(),
        m_strClientData=reader.read_string(),
    )


@dataclass
class CRespGameClientData:
    m_stRetMsg: CCommonRespMsg
    m_strClientData: str = ""  # String


def write_c_resp_game_client_data(
    writer: BinaryWriter, resp: CRespGameClientData
) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_string(resp.m_strClientData)


# =============================================================================
# PVE SEASON
# =============================================================================


@dataclass
class CPveSeasonInfo:
    m_nType: int  # UInt16
    m_nSeasonId: int  # UInt32
    m_nTransID: int  # UInt32


def read_c_pve_season_info(reader: BinaryReader) -> CPveSeasonInfo:
    return CPveSeasonInfo(
        m_nType=reader.read_uint16(),
        m_nSeasonId=reader.read_uint32(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class CRespPveSeasonInfo:
    m_stRetMsg: CCommonRespMsg
    m_nSeasonId: int = 0  # UInt32
    m_nRank: int = 0  # UInt32
    m_nScore: int = 0  # UInt32


def write_c_resp_pve_season_info(
    writer: BinaryWriter, resp: CRespPveSeasonInfo
) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint32(resp.m_nSeasonId)
    writer.write_uint32(resp.m_nRank)
    writer.write_uint32(resp.m_nScore)


# =============================================================================
# FISHING
# =============================================================================


@dataclass
class CReqGameFishing:
    m_nType: int  # UInt16
    m_nTransID: int  # UInt32


def read_c_req_game_fishing(reader: BinaryReader) -> CReqGameFishing:
    return CReqGameFishing(
        m_nType=reader.read_uint16(),
        m_nTransID=reader.read_uint32(),
    )


@dataclass
class STGameFishingRank:
    m_nRank: int  # UInt32
    m_nScore: int  # UInt32
    m_strName: str  # String


def write_st_game_fishing_rank(writer: BinaryWriter, rank: STGameFishingRank) -> None:
    writer.write_uint32(rank.m_nRank)
    writer.write_uint32(rank.m_nScore)
    writer.write_string(rank.m_strName)


@dataclass
class CRespGameFishing:
    m_stRetMsg: CCommonRespMsg
    m_nScore: int = 0  # UInt32
    m_vecRanks: List[STGameFishingRank] = field(default_factory=list)


def write_c_resp_game_fishing(writer: BinaryWriter, resp: CRespGameFishing) -> None:
    write_c_common_resp_msg(writer, resp.m_stRetMsg)
    writer.write_uint32(resp.m_nScore)
    writer.write_array(resp.m_vecRanks, lambda r: write_st_game_fishing_rank(writer, r))
